### PR TITLE
[2201.1.x] Added new endpoint to get function definition given position

### PR DESCRIPTION
**Pull request by axewilledge**
Fixes # https://github.com/wso2-enterprise/choreo/issues/12425
https://github.com/ballerina-platform/ballerina-lang/issues/37414

## Check List 
- [ ] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [ ] Added necessary tests
  - [x] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [x] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/extensions/ballerina/document/BallerinaDocumentService.java [hunk 1]
```diff
@@ -289,14 +295,103 @@ public CompletableFuture<BallerinaSyntaxTreeResponse> triggerModify(BallerinaTri
...
+            BallerinaSyntaxTreeResponse reply = new BallerinaSyntaxTreeResponse();
+            String fileUri = request.getDocumentIdentifier().getUri();
+            Optional<Path> filePath = PathUtil.getPathFromURI(fileUri);
+
+            if (filePath.isEmpty()) {
+                return reply;
+            }
+
+            try {
+                Optional<Document> srcFile = this.workspaceManagerProxy.get().document(filePath.get());
```
Comment (IMS94) [inline language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/extensions/ballerina/document/BallerinaDocumentService.java:310]: Is this something going to be called from the expression editor? If that's the case, the file URI will have `expr` scheme. We should use the `workspaceManegerProxy.get(fileUri)` method then.

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/extensions/ballerina/document/BallerinaDocumentService.java [hunk 2]
```diff
@@ -289,14 +295,103 @@ public CompletableFuture<BallerinaSyntaxTreeResponse> triggerModify(BallerinaTri
...
+                if (srcFile.isEmpty()) {
+                    return reply;
+                }
+
+                // Get the semantic model.
+                Optional<SemanticModel> semanticModel = this.workspaceManagerProxy.get().semanticModel(filePath.get());
```
Comment (IMS94) [inline language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/extensions/ballerina/document/BallerinaDocumentService.java:316]: Same as above. Use the `get(fileUri)` method if this is called from the diagram editor/expression editor.

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/extensions/ballerina/document/BallerinaDocumentService.java [hunk 3]
```diff
@@ -289,14 +295,103 @@ public CompletableFuture<BallerinaSyntaxTreeResponse> triggerModify(BallerinaTri
...
+                        LinePosition.from(lineValue, charValue));
+                if (functionSymbol.isEmpty()) {
+                    return reply;
+                }
+
+                // Get the file path of the function symbol
+                String functionPath = functionSymbol.get().getLocation().get().lineRange().filePath();
+
+                // Get the project of current file
+                Optional<Project> project = this.workspaceManagerProxy.get().project(filePath.get());
```
Comment (IMS94) [inline language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/extensions/ballerina/document/BallerinaDocumentService.java:333]: Same as above

**Review by IMS94 — CHANGES_REQUESTED** [review]

**Review by NipunaRanasinghe — APPROVED** [review]
LGTM

**Review by IMS94 — APPROVED** [review]

**Pull request merged**